tests/FileStorageAbsTest.php

```diff
@@ -0,0 +1,138 @@
+    ];
+
+    public function testFileStorageAbsCreate()
+    {
+        $storage = $this->client->createAbsFileStorage(self::DEFAULT_ABS_OPTIONS);
+
+        $this->assertArrayNotHasKey('accountKey', $storage);
+        $this->assertSame($storage['accountName'], TEST_ABS_ACCOUNT_NAME);
+        $this->assertSame($storage['containerName'], TEST_ABS_CONTAINER_NAME);
+        $this->assertSame($storage['provider'], 'azure');
+        $this->assertSame($storage['isDefault'], false);
+    }
+
+    public function testRotateAbsKey()
+    {
+        $storage = $this->client->createAbsFileStorage(self::DEFAULT_ABS_OPTIONS);
+
+        $rotatedStorage = $this->client->rotateAbsFileStorageCredentials($storage['id'], self::ROTATE_ABS_OPTIONS);
+        $this->assertArrayNotHasKey('accountKey', $storage);
+        $this->assertSame($rotatedStorage['accountName'], TEST_ABS_ACCOUNT_NAME);
+        $this->assertSame($rotatedStorage['containerName'], TEST_ABS_CONTAINER_NAME);
+        $this->assertSame($rotatedStorage['provider'], 'azure');
+        $this->assertSame($rotatedStorage['isDefault'], false);
+    }
+
+    public function testListAbsStorages()
+    {
+        $initCount = count($this->client->listAbsFileStorage());
+        $this->client->createAbsFileStorage(self::DEFAULT_ABS_OPTIONS);
+        $storages = $this->client->listAbsFileStorage();
+
+        $this->assertSame($initCount + 1, count($storages));
+
+        foreach ($storages as $storage) {
+            if ($storage['provider'] !== 'azure') {
+                $this->fail('List of Azure Blob Storages contains also S3 Storage');
+            }
+        }
+    }
+
+    public function testSetAbsStorageAsDefault()
+    {
+        $storage = $this->client->createAbsFileStorage(self::DEFAULT_ABS_OPTIONS);
+
+        $this->assertFalse($storage['isDefault']);
+
+        $storage = $this->client->setAbsFileStorageAsDefault($storage['id']);
+
+        $this->assertTrue($storage['isDefault']);
+
+        $storageList = $this->client->listAbsFileStorage();
+        foreach ($storageList as $item) {
+            if ($item['isDefault'] && $item['id'] !== $storage['id']) {
+                $this->fail('There are more storage with default flag');
+            }
+        }
+    }
+
+    public function testCrossProviderStorageDefaultAbsS3()
+    {
+        $storage = $this->client->createAbsFileStorage(self::DEFAULT_ABS_OPTIONS);
+
+        $this->expectException(ClientException::class);
+        $this->expectExceptionMessage(sprintf('AWS S3 file storage "%d" not found', $storage['id']));
+        $this->client->setS3FileStorageAsDefault($storage['id']);
+    }
+
+    public function testCreateAbsStorageWithoutRequiredParam()
+    {
+        $this->expectException(ClientException::class);
+        $this->expectExceptionMessage('"accountName": This field is missing. "accountKey": This field is missing. "containerName": This field is missing. "owner": This field is missing.');
```

Comment:
Přesunu sem vlákno z https://github.com/keboola/kbc-manage-api-php-client/pull/76#discussion_r294683852

Mně se tohle furt hrozně nelíbí. Exception message by měla bejt třeba jen něco jako `Bad request` a tyhle strukturovaný data v tom potenciálním novým fieldu.

Reply:
Exception message aktuálně máme například "Missing required parameters A,B,C". Tohle mi přijde jako ekvivalent.

Ale jak jsem psal někde na žacátku, tak tohle se bude používat hlavně pak na UI, takže jestli mi hodíš JSON příklad, jak si představuješ, že to má přesně vypadat tak já to do tý podoby předělám. Jediný v čem nechci ustoupit je vzdát se Symfony validátoru a dělat `array_diff` na required params.

Reply:
nechci tady z oleje a bez kontextu vymejšlet, jak maj vypadat chyby v JSONu. s validátorem samozřejmě souhlas. 

když to řeknu hodně zjednodušeně - proč se vůbec zabejváme chybovou hláškou něčeho, co používáme jen my a jen přes api (bez ui)? v `error` by mohlo bejt `Bad request` a v `errors` v podstatě jakýkoliv strukturovaný data o všech chybách. jediný, co bych řekl, že by bylo pěkný místo

```json
{
    "errors": {
        "field1": "Missing value"
    }
}
```

použít

```json
{
    "errors": [
        {
            "key": "field1",
            "message": "Missing value"
        }
    ]
}

protože je to asi tak milionkrát flexibilnější. při práci s API to bude takhle i krásně čitelný místo jedný megadlouhý věty, kterou si budeš muset v hlavě parsovat do jednotlivejch errorů

Reply:
Udělám tedy nový field `errors`, který bude přebírat rovnou `ConstraintViolationInterface []`, aby jsme s tím měli co nejmíň práce. A podhodíš tomu rovnou chyby formuláře. Tohle mi trochu příjde na samostatný PR, ale zase bych to neměl na čem vyzkoušet...

Reply:
Aktuální error response vypadá takhle:
```json

Reply:
Chybovka klienta pak vypadá takhle:
```
Provided data are not valid.
Errors:
"accountName": This field is missing.
"accountKey": This field is missing.
"containerName": This field is missing.
"owner": This field is missing.
'
```
To je už jenom o formátování. Neměl by být problém s kompatibilitou se starou verzí vyjímek.

Reply:
kdyžtak `Provided data is not valid.` ale imho i to je špatně, neposíláš mu žádný data, odesíláš request. Nestačilo by fakt jen `Invalid request`?

Reply:
Dal jsem tam `Invalid request`